silkworm/sync/sentry_client.cpp

```diff
@@ -16,137 +16,300 @@
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_by_min_block(std::move(message), max_peers));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+void SentryClient::peer_min_block(const PeerId& peer_id, BlockNum /*min_block*/) {
+    auto peer_public_key = sentry::common::EccPublicKey::deserialize(peer_id);
+    sync_spawn(tasks_, io_context_, sentry_client_->service()->peer_min_block(std::move(peer_public_key)));
+}
 
+void SentryClient::execution_loop() {
     log::set_thread_name("sentry-recv   ");
 
+    auto tasks_completion = [](const std::exception_ptr& ex_ptr) {
+        try {
+            if (ex_ptr) std::rethrow_exception(ex_ptr);
+        } catch (const boost::system::system_error& ex) {
+            if (ex.code() == boost::system::errc::operation_canceled) {
+                log::Debug(kLogTitle) << "execution_loop tasks.wait completed";
+            } else {
+                log::Error(kLogTitle) << "execution_loop tasks.wait system_error: " << ex.what();
+                throw;
+            }
+        } catch (...) {
+            log::Error(kLogTitle) << "execution_loop tasks.wait unexpected exception";
+            throw;
+        }
+    };
+    co_spawn(io_context_, tasks_.wait(), bind_cancellation_slot(tasks_cancellation_signal_.slot(), tasks_completion));
+
     while (!is_stopping()) {
         try {
             connected_ = false;
-            log::Info("SentryClient", {"remote", sentry_addr_}) << " connecting ...";
+            log::Info(kLogTitle) << "connecting ...";
 
             // send current status of the chain
-            hand_shake();
+            handshake();
             set_status();
 
             connected_ = true;
             connected_.notify_all();
-            log::Info("SentryClient", {"remote", sentry_addr_}) << " connected";
-
-            // send a message subscription
-            auto rpc = std::make_shared<rpc::ReceiveMessages>(RMS::BlockAnnouncements | RMS::BlockRequests);
-            std::atomic_store(&receive_messages_, rpc);
-
-            exec_remotely(*receive_messages_);
+            log::Info(kLogTitle) << "connected";
 
             // receive messages
-            while (!is_stopping() && receive_messages_->receive_one_reply()) {
-                const auto& message = receive_messages_->reply();
 
-                publish(message);
-            }
+            std::function<awaitable<void>(silkworm::sentry::api::api_common::MessageFromPeer)> consumer = [this](auto message_from_peer) -> awaitable<void> {
+                if (!this->is_stopping()) {
+                    co_await this->publish(message_from_peer);
+                }
+            };
+
+            sync_spawn(tasks_, io_context_, sentry_client_->service()->messages(make_message_id_filter(), consumer));
```

Comment:
I would like to rename the sentry client's `messages()` method to `receive_messages()` : with this modification, this code will be more readable.

Reply:
I understand. The idea was that the names exactly match to sentry.proto, but I see how it is confusing having this callback style of subscribing. I wish I could just return a lazy "stream" of messages, but it was tricker than I thought.

Thinking about a different name, we have another method like "node_info", but we are not calling it "receive_node_info". Maybe "wait_for_messages" or "on_messages"? I need to think a bit more before renaming this. Same for peer_events e.g. "on_peer_events" or "subscribe_to_peer_events_and_wait" 🤔